i2c/bmp280_i2c/bmp280_i2c.c

```diff
@@ -0,0 +1,260 @@
+const uint8_t REG_DIG_P6_MSB = 0x99;
+const uint8_t REG_DIG_P7_LSB = 0x9A;
+const uint8_t REG_DIG_P7_MSB = 0x9B;
+const uint8_t REG_DIG_P8_LSB = 0x9C;
+const uint8_t REG_DIG_P8_MSB = 0x9D;
+const uint8_t REG_DIG_P9_LSB = 0x9E;
+const uint8_t REG_DIG_P9_MSB = 0x9F;
+
+const uint8_t BMP280_WRITE_MODE = 0xFE;
+const uint8_t BMP280_READ_MODE = 0xFF;
+
+struct bmp280_calib_param {
+  // temperature params
+  uint16_t dig_t1;
+  int16_t dig_t2;
+  int16_t dig_t3;
+
+  // pressure params
+  uint16_t dig_p1;
+  int16_t dig_p2;
+  int16_t dig_p3;
+  int16_t dig_p4;
+  int16_t dig_p5;
+  int16_t dig_p6;
+  int16_t dig_p7;
+  int16_t dig_p8;
+  int16_t dig_p9;
+
+  // intermediate result for pressure calc
+  int32_t t_fine;
+};
+
+#ifdef i2c_default
+void bmp280_init() {
+  // use the "handheld device dynamic" optimal setting (see datasheet)
+  uint8_t buf[2];
+
+  // 500ms sampling time, x16 filter, not set
+  uint8_t reg_config_val = ((0x04 << 5) | (0x05 << 2)) & 0xFC;
+
+  // LSB of slave address sets read or write mode
+  // send register number followed by its corresponding value
+  buf[0] = REG_CONFIG;
+  buf[1] = reg_config_val;
+  i2c_write_blocking(i2c_default, (addr & BMP280_WRITE_MODE), buf, 2, false);
+
+  // osrs_t x1, osrs_p x4, normal mode operation
+  uint8_t reg_ctrl_meas_val = (0x01 << 5) | (0x03 << 2) | (0x03);
+  buf[0] = REG_CTRL_MEAS;
+  buf[1] = reg_ctrl_meas_val;
+  i2c_write_blocking(i2c_default, (addr & BMP280_WRITE_MODE), buf, 2, false);
+}
+
+void bmp280_read_raw(int32_t *temp, int32_t *pressure) {
+  // BMP280 data registers are auto-incrementing and we have 3 temperature and
+  // pressure registers, each so we start at 0xF7 and read 6 bytes to 0xFC note:
+  // normal mode does not require further ctrl_meas and config register writes
+
+  uint8_t buf[6];
+  i2c_write_blocking(i2c_default, (addr & BMP280_WRITE_MODE), &REG_PRESSURE_MSB,
+                     1, true);  // true to keep master control of bus
+  i2c_read_blocking(i2c_default, (addr & BMP280_READ_MODE), buf, 6,
+                    false);  // false - finished with bus
+
+  // store the 20 bit read in a 32 bit signed integer for conversion
+  *pressure = (buf[0] << 12) | (buf[1] << 4) | (buf[2] >> 4);
+  *temp = (buf[3] << 12) | (buf[4] << 4) | (buf[5] >> 4);
+}
+
+void bmp280_reset() {
+  // reset the device with the power-on-reset procedure
+  uint8_t buf[2] = {REG_RESET, 0xB6};
+  i2c_write_blocking(i2c_default, (addr & BMP280_WRITE_MODE), buf, 2, false);
+}
+
+float bmp280_convert_temp(int32_t *temp, struct bmp280_calib_param *params) {
+  // use the 32-bit floating point compensation implementation given in the
+  // datasheet
+  float var1, var2;
+  var1 = (((float)*temp) / 16384.0 - ((float)params->dig_t1) / 1024.0) *
+         ((float)params->dig_t2);
+  var2 = ((((float)*temp) / 131072.0 - ((float)params->dig_t1) / 8192.0) *
+          (((float)*temp) / 131072.0 - ((float)params->dig_t1) / 8192.0)) *
+         ((float)params->dig_t3);
```

Comment:
You seem to be calculating `(((float)*temp) / 131072.0 - ((float)params->dig_t1) / 8192.0)` twice here - is that correct? :shrug:

Reply:
Yes, it seem so. The floating point implementation from Bosch had this exact line. I've swapped it for the fixed point so this doesn't exist anymore anyway.